[2.x] NoClassDefFoundError: $Wrap

## steps

1. Load a build.
2. Change build.sbt
3. Reload

## problem

The build crashes with `NoClassDefFoundError: $Wrap<hex>`

```scala
[info] welcome to sbt 2.0.0-alpha8-SNAPSHOT (Azul Systems, Inc. Java 1.8.0_352)
[info] loading project definition from /private/tmp/foo/project
[info] loading settings for project foo from build.sbt ...
[info] loading settings for project foo from build.sbt ...
java.lang.NoClassDefFoundError: $Wrape8743d4f36$
	at $Wrap0b8ea34d40$.$anonfun$1(build.sbt:1)
	at sbt.internal.util.EvaluateSettings.constant$$anonfun$1(INode.scala:213)
	at sbt.internal.util.EvaluateSettings$MixedNode.evaluate0(INode.scala:230)
	at sbt.internal.util.EvaluateSettings$INode.evaluate(INode.scala:172)
	at sbt.internal.util.EvaluateSettings.submitEvaluate$$anonfun$1(INode.scala:93)
	at sbt.internal.util.EvaluateSettings.submitEvaluate$$anonfun$adapted$1(INode.scala:93)
	at scala.Function0.apply$mcV$sp(Function0.scala:42)
	at sbt.internal.util.EvaluateSettings.run0(INode.scala:105)
	at sbt.internal.util.EvaluateSettings.submit$$anonfun$1(INode.scala:101)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.ClassNotFoundException: $Wrape8743d4f36$
	at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
	... 12 more
[error] java.lang.NoClassDefFoundError: $Wrape8743d4f36$
[error] Use 'last' for the full log.
[warn] Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore? (default: r)
```

## expectation

The build reloads.

## notes

I am guessing that there's a bug in invalidation logic somewhere.
Current workaround is to remove `project/target` directory when this happens.

--- a/buildfile/src/main/scala/sbt/internal/Eval.scala
+++ b/buildfile/src/main/scala/sbt/internal/Eval.scala
@@ -216,6 +216,10 @@ class Eval(
     digester.update(bytes(ev.extraHash))
     // Include SNAPSHOT classpath hash to invalidate cache when sbt version changes (fixes #7713)
     digester.update(bytes(snapshotClasspathHash))
+    // Include imports in hash to invalidate cache when definition module names change (fixes #7424)
+    imports.strings.foreach { imp =>
+      digester.update(bytes(imp))
+    }
     val d = digester.digest()
     val hash = Hash.toHex(d)
     val moduleName = makeModuleName(hash)

--- a/main/src/main/scala/sbt/internal/Load.scala
+++ b/main/src/main/scala/sbt/internal/Load.scala
@@ -866,11 +866,9 @@ private[sbt] object Load {
           )
         }
       val loadedProjects = processAutoAggregate(loadedProjects0, uri)
-      // TODO: Uncomment when we fixed https://github.com/sbt/sbt/issues/7424
-      // likely keepClassFiles isn't covering enough.
-      // timed("Load.loadUnit: cleanEvalClasses", log) {
-      //   cleanEvalClasses(defDir, keepClassFiles)
-      // }
+      timed("Load.loadUnit: cleanEvalClasses", log) {
+        cleanEvalClasses(defDir, keepClassFiles)
+      }
       val defs = if (defsScala.isEmpty) defaultBuildIfNone :: Nil else defsScala
       // HERE we pull out the defined vals from memoSettings and unify them all so
       // we can use them later.

--- a/sbt-app/src/sbt-test/project/reload-val-change/build.sbt
+++ b/sbt-app/src/sbt-test/project/reload-val-change/build.sbt
@@ -1,0 +1,2 @@
+val myVersion = "1.0.0"
+name := myVersion

--- a/sbt-app/src/sbt-test/project/reload-val-change/changes/build.sbt
+++ b/sbt-app/src/sbt-test/project/reload-val-change/changes/build.sbt
@@ -1,0 +1,2 @@
+val myVersion = "2.0.0"
+name := myVersion

--- a/sbt-app/src/sbt-test/project/reload-val-change/test
+++ b/sbt-app/src/sbt-test/project/reload-val-change/test
@@ -1,0 +1,5 @@
+# https://github.com/sbt/sbt/issues/7424
+> name
+$ copy-file changes/build.sbt build.sbt
+> reload
+> name

